File: potato-mesher/python/genboxes.py
#!/bin/python

# doos(driehoeken):

#     randen = bereken randen (driehoeken)
#     linkerdriehoeken = bereken linkerdriehoeken (driehoeken)
#     rechterdriehoeken = bereken rechterdriehoeken (driehoeken)

#     als # linkerdriehoeken > N:
#         linkerdoos = doos(linkerdriehoeken)
#     anders
#         linkerdoos = linkerdriehoeken

#     als # rechterdriehoeken > N:
#         rechterdoos = doos(rechterdriehoeken)
#     anders:
#         rechterdoos = rechterdriehoeken

#     return (randen, linkerdoos, rechterdoos)


# rootdoos = doos(alle driehoeken)

import armadillo
import sys
import operator

N = 10

def find_minmax_vertices(triangles):
    min_x, min_y, min_z = sys.maxsize, sys.maxsize, sys.maxsize
    max_x, max_y, max_z = -min_x, -min_y, -min_z
    for triangle in triangles:
        xs = [armadillo.vertices[vi][0] for vi in triangle]
        ys = [armadillo.vertices[vi][1] for vi in triangle]
        zs = [armadillo.vertices[vi][2] for vi in triangle]
        max_x = max([max_x] + xs)
        max_y = max([max_y] + ys)
        max_z = max([max_z] + zs)
        min_x = min([min_x] + xs)
        min_y = min([min_y] + ys)
        min_z = min([min_z] + zs)

    bv1 = (max_x, max_y, max_z)
    bv2 = (min_x, min_y, min_z)
    return bv1, bv2

def divide_box(boxvertex1, boxvertex2):
    print(f"dividing {boxvertex1} and {boxvertex2}")

    x1, y1, z1 = boxvertex1
    x2, y2, z2 = boxvertex2
    x = ('x', abs(x1 - x2))
    y = ('y', abs(y1 - y2))
    z = ('z', abs(z1 - z2))

    print(x)
    print(y)
    print(z)

    longest = max(x, y, z, key=operator.itemgetter(1))
    x = x[1] if longest[0] != 'x' else longest[1] / 2
    y = y[1] if longest[0] != 'y' else longest[1] / 2
    z = z[1] if longest[0] != 'z' else longest[1] / 2

    emptyleftbox = (boxvertex1, (
        x1 - x,
        y1 - y,
        z1 - z
    ))
    emptyrightbox = ((
        x2 + x,
        y2 + y,
        z2 + z
    ), boxvertex2)

    return emptyleftbox, emptyrightbox

def find_inside(triangles, box):
    n = len(triangles)
    print(f"finding {n} triangles in {box}")

    (x1, y1, z1), (x2, y2, z2) = box
    for triangle in triangles:
        xs = [armadillo.vertices[vi][0] for vi in triangle]
        ys = [armadillo.vertices[vi][1] for vi in triangle]
        zs = [armadillo.vertices[vi][2] for vi in triangle]
        inside = x1 >= max(xs) \
            and x2 <= min(xs) \
            and y1 >= max(ys) \
            and y2 <= min(ys) \
            and z1 >= max(zs) \
            and z2 <= min(zs)
        if(inside):
            yield triangle

def box(triangles):

    n = len(triangles)
    print(f"Starting box with: {n} triangles")

    boxvertex1, boxvertex2 = find_minmax_vertices(triangles)
    emptyleftbox, emptyrightbox = divide_box(boxvertex1, boxvertex2)
    left_triangles = list(find_inside(triangles, emptyleftbox))
    right_triangles = list(find_inside(triangles, emptyrightbox))

    n = len(left_triangles)
    print(f"left: {n} triangles")
    n = len(right_triangles)
    print(f"right: {n} triangles")

    if len(left_triangles) > N:
        leftbox = box(left_triangles)
    else:
        leftbox = left_triangles

    if len(right_triangles) > N:
        rightbox = box(right_triangles)
    else:
        rightbox = right_triangles

    return (boxvertex1, boxvertex2, leftbox, rightbox)

# TODO
# def with_triangle_data(box):
#     v1, v2, leftbox, rightbox = box
#     if(all([isinstance(x, int) for x in leftbox])):

File: potato-mesher/clojure/genboxes.clj
(defn parse-mesh [filename]
  (let [[n-of-vertices-str & tail] (->> (slurp filename) clojure.string/split-lines)
        n-of-vertices (read-string n-of-vertices-str)
        [vertices newtail] (split-at n-of-vertices tail)
        [_ & triangles] newtail
        strs-to-nums (fn [strs] (map (fn [s] (read-string (str "(" s ")"))) strs))
        vertices (strs-to-nums vertices)
        triangles (strs-to-nums triangles)]
    (map (fn [triangle]
           (let [[vi1, vi2, vi3] triangle]
             [(nth vertices vi1)
              (nth vertices vi2)
              (nth vertices vi3)
              ]))
         triangles)))

; (defn to-file [path form]
;   (with-open [w (java.io.FileWriter. (clojure.java.io/file path))]
;     (print-dup form w)))

; (defn from-file [path]
;   (with-open [r (java.io.PushbackReader. (java.io.FileReader. (clojure.java.io/file path)))]
;      (read r)))

; Pseudo:
;
; doos(driehoeken):
;
;     randen = bereken randen (driehoeken)
;     linkerdriehoeken = bereken linkerdriehoeken (driehoeken)
;     rechterdriehoeken = bereken rechterdriehoeken (driehoeken)
;
;     als # linkerdriehoeken > N:
;         linkerdoos = doos(linkerdriehoeken)
;     anders
;         linkerdoos = linkerdriehoeken
;
;     als # rechterdriehoeken > N:
;         rechterdoos = doos(rechterdriehoeken)
;     anders:
;         rechterdoos = rechterdriehoeken
;
;     return (randen, linkerdoos, rechterdoos)
;
;
; rootdoos = doos(alle driehoeken)

(def DEBUG true)

; (defn find-minmax-vertices [triangles]
;  (->> triangles
;       (apply concat) ; flatten triangles to one big vertex list
;       (apply map vector) ; pivot vertexs to xs, ys, zs
;       (map (fn [coords]
;              [(apply max coords) (apply min coords)]))
;       (apply map vector))) ;pivot minmax vals to back to vertexes
(defn minmax [[[sminx sminy sminz] [smaxx smaxy smaxz]] [nxx nxy nxz]]
  (list
    (list (min sminx nxx)
          (min sminy nxy)
          (min sminz nxz))
    (list (max smaxx nxx)
          (max smaxy nxy)
          (max smaxz nxz)
          )))

(defn find-minmax-vertices [triangles]
  (let [[head & tail] (apply concat triangles)]
    (reduce minmax [head head] tail)))

(defn divide-box [boxvertex1 boxvertex2]
  (if DEBUG (print "dividing " boxvertex1 " and " boxvertex2))

  (let [[x1 y1 z1] boxvertex1
        [x2 y2 z2] boxvertex2
        x (Math/abs (- x1 x2))
        y (Math/abs (- y1 y2))
        z (Math/abs (- z1 z2))
        [longest & others] (->> [["x" x] ["y" y] ["z" z]] (sort-by second) reverse)
        halved [(first longest) (/ (second longest) 2)]
        [x y z] (->> (conj others halved) sort (map second))
        emptyleftbox [boxvertex1 [(- x1 x)
                                  (- y1 y)
                                  (- z1 z)]]
        emptyrightbox [[(+ x2 x)
                        (+ y2 y)
                        (+ z2 z)] boxvertex2]]

        [emptyleftbox emptyrightbox]
        ))

(defn find-completly-inside [triangles box]
  (if DEBUG (print "find inside: " (type box)))

  (let [[[x1 y1 z1] [x2 y2 z2]] box]
    (filter (fn [triangle]
              (let [[xs ys zs] (apply map vector triangle)]
                (every? identity [(>= x1 (apply max xs))
                                  (<= x2 (apply min xs))
                                  (>= y1 (apply max ys))
                                  (<= y2 (apply min ys))
                                  (>= z1 (apply max zs))
                                  (<= z2 (apply min zs))]
                        )))
            triangles)))

; (defn makebox [triangles]
;   (if DEBUG (print "triangles: " (count triangles)))

;   (let [[boxvertex1 boxvertex2] (find-minmax-vertices triangles)
;         [emptyleftbox emptyrightbox] (divide-box boxvertex1 boxvertex2)
;         left-triangles (find-completly-inside triangles emptyleftbox)
;         right-triangles (find-completly-inside triangles emptyrightbox)

;         leftbox (cond (> (count left-triangles) N) (makebox left-triangles)
;                       :else left-triangles)
;         rightbox (cond (> (count right-triangles) N) (makebox right-triangles)
;                       :else right-triangles)]
;     {:boxvertices [boxvertex1 boxvertex2]
;      :leftbox leftbox
;      :rightbox rightbox}))

(defn makeboxV2 [N triangles]
  (if DEBUG (print "triangles: " (count triangles)))

  (let [[boxvertex1 boxvertex2] (find-minmax-vertices triangles)
        [emptyleftbox emptyrightbox] (divide-box boxvertex1 boxvertex2)
        left-triangles (find-completly-inside triangles emptyleftbox)
        right-triangles (find-completly-inside triangles emptyrightbox)
        leftover-triangles (clojure.set/difference (set triangles) (set (concat left-triangles right-triangles)))
        [firsthalf secondhalf] (split-at (/ (count leftover-triangles) 2) leftover-triangles)
        left-triangles (concat left-triangles firsthalf)
        right-triangles (concat right-triangles secondhalf)

        leftbox (cond (> (count left-triangles) N) (makeboxV2 N left-triangles)
                      :else left-triangles)
        rightbox (cond (> (count right-triangles) N) (makeboxV2 N right-triangles)
                      :else right-triangles)]
    {:boxvertices [boxvertex1 boxvertex2]
     :leftbox leftbox
     :rightbox rightbox}))

(defn triangle-writer! [writer triangles]
    (doseq [t triangles]
      (.write writer (str "triangle " (clojure.string/join " " (flatten t)) "\n")))
    (.write writer (str (if (empty? triangles) "emptybox" "box") "\n")))

(defn box-writer! [writer box]
  (let [lb (box :leftbox)
        rb (box :rightbox)
        lb-isbox (= (type lb) clojure.lang.PersistentArrayMap)
        rb-isbox (= (type rb) clojure.lang.PersistentArrayMap)]
  (if lb-isbox
    (box-writer! writer lb)
    (triangle-writer! writer lb))
  (if rb-isbox
    (box-writer! writer rb )
    (triangle-writer! writer rb))
  (.write writer "box\n")))

(defn write-optimized-mesh-file! [from to N]
  (with-open [writer (clojure.java.io/writer to)]
      (->> (parse-mesh from) (makeboxV2 N) (box-writer! writer))))
